src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs

```diff
@@ -461,24 +467,41 @@ public class PatchDocumentCommand : PatchDocumentCommandBase
             _ignoreMaxStepsForScript = ignoreMaxStepsForScript;
             if (string.IsNullOrEmpty(id) || id.EndsWith(identityPartsSeparator) || id.EndsWith('|'))
                 throw new ArgumentException($"The ID argument has invalid value: '{id}'", nameof(id));
+
+            // If the caller is a DocumentsOperationContext, then we can apply the optimization.
+            if (context is DocumentsOperationContext doContext)
+                InitializeCmd(doContext);
         }
 
-        protected override long ExecuteCmd(DocumentsOperationContext context)
+        private void InitializeCmd(DocumentsOperationContext context)
         {
-            ScriptRunner.SingleRun runIfMissing = null;
             _database = context.DocumentDatabase;
+            _returnRun = _database.Scripts.GetScriptRunner(_patch.Run, readOnly: false, out _run);
+            _disposableStatement = _ignoreMaxStepsForScript ? _run.ScriptEngine.DisableMaxStatements() : null;
+            _disposableScriptRunner = _patchIfMissing.Run != null ? _database.Scripts.GetScriptRunner(_patchIfMissing.Run, readOnly: false, out _runIfMissing) : null;
+        }
 
-            using (_database.Scripts.GetScriptRunner(_patch.Run, readOnly: false, out var run))
-            {
-                using (_ignoreMaxStepsForScript ? run.ScriptEngine.DisableMaxStatements() : null)
-                {
-                    using (_patchIfMissing.Run != null ? _database.Scripts.GetScriptRunner(_patchIfMissing.Run, readOnly: false, out runIfMissing) : (IDisposable)null)
-                    {
-                        PatchResult = ExecuteOnDocument(context, _id, _expectedChangeVector, run, runIfMissing);
-                        return 1;
-                    }
-                }
-            }
+        protected override long ExecuteCmd(DocumentsOperationContext context)
+        {
+            // PERF: Since we are not able to apply the optimization of shifting the cost of this operations into the constructor
+            // we will do it here instead. 
+            if (_database == null)
```

Comment:
`InitializeCmd` is called previous if:

```
if (context is DocumentsOperationContext doContext)
```

Why is the check here different than the previous one?

Reply:
Can we normalize the check? Add a `IsInit` or something like that flag? Would rather avoid two separate checks

Reply:
Done